<a href="https://colab.research.google.com/github/Z3R0-2X/Statistical-Learning-e22122/blob/main/Assignment_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


### Part A: Theoretical Fundamentals
1. Derivation of the Bias in the MLE Covariance Matrix

We begin with the Maximum Likelihood Estimator for the population covariance matrix $\boldsymbol{\Sigma}$:


$$\widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{1}{n} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T$$


where $\widehat{\boldsymbol{\mu}}_n = \frac{1}{n}\sum_{i=1}^n \mathbf{X}_i$ is the empirical sample mean.

To determine if this estimator is biased, we expand the deviations around the true, hidden population mean $\boldsymbol{\mu}$:


$$\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n = (\mathbf{X}_i - \boldsymbol{\mu}) - (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})$$

Substituting this back into the sum inside the MLE formula gives:


$$\sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T = \sum_{i=1}^n \left[ (\mathbf{X}_i - \boldsymbol{\mu}) - (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \right] \left[ (\mathbf{X}_i - \boldsymbol{\mu}) - (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \right]^T$$

Expanding the product matrix term-by-term yields:


$$= \sum_{i=1}^n (\mathbf{x}_i - \boldsymbol{\mu})(\mathbf{x}_i - \boldsymbol{\mu})^T - \sum_{i=1}^n (\mathbf{x}_i - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T - \sum_{i=1}^n (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\mathbf{x}_i - \boldsymbol{\mu})^T + \sum_{i=1}^n (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T$$

Notice that $\sum_{i=1}^n (\mathbf{X}_i - \boldsymbol{\mu}) = n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})$. Using this simplification, the two middle cross-product terms contract:


$$= \sum_{i=1}^n (\mathbf{X}_i - \boldsymbol{\mu})(\mathbf{X}_i - \boldsymbol{\mu})^T - n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T - n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T + n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T$$

$$= \sum_{i=1}^n (\mathbf{X}_i - \boldsymbol{\mu})(\mathbf{X}_i - \boldsymbol{\mu})^T - n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T$$

Now, we apply the expectation operator $\mathbb{E}[\cdot]$ to both sides:


$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{1}{n} \sum_{i=1}^n \mathbb{E}[(\mathbf{X}_i - \boldsymbol{\mu})(\mathbf{X}_i - \boldsymbol{\mu})^T] - \mathbb{E}[(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T]$$

By definition of the structural population moments:

* $\mathbb{E}[(\mathbf{X}_i - \boldsymbol{\mu})(\mathbf{X}_i - \boldsymbol{\mu})^T] = \boldsymbol{\Sigma}$
* $\mathbb{E}[(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T] = \text{Var}(\widehat{\boldsymbol{\mu}}_n) = \frac{1}{n}\boldsymbol{\Sigma}$

Substituting these back into our expectation equation gives:


$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{1}{n} (n\boldsymbol{\Sigma}) - \frac{1}{n}\boldsymbol{\Sigma} = \boldsymbol{\Sigma} - \frac{1}{n}\boldsymbol{\Sigma} = \left(\frac{n-1}{n}\right)\boldsymbol{\Sigma}$$

Because $\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] \neq \boldsymbol{\Sigma}$, the estimator is systematically downward biased. This occurs because calculating variations against the *sample* mean rather than the *true* population mean artificially underrepresents the real spatial scattering of the system.

**Bessel's Correction Elimination of Bias**

To yield an unbiased estimator ($\mathbf{S}$), we scale the maximum likelihood matrix by the reciprocal of the bias factor ($\frac{n}{n-1}$):


$$\mathbf{S} = \frac{n}{n-1} \widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{n}{n-1} \left( \frac{1}{n} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T \right) = \frac{1}{n-1} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T$$

Evaluating the expectation confirms its lack of bias:


$$\mathbb{E}[\mathbf{S}] = \frac{n}{n-1} \mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{n}{n-1} \left(\frac{n-1}{n}\right)\boldsymbol{\Sigma} = \boldsymbol{\Sigma}$$

---

**Operational Error Diagnostics**

* **Type I Error ($\alpha$):** A **False Alarm**. This happens when the diagnostic system concludes that the structure is damaged or unstable (rejects $H_0$), when in reality the physical system remains completely nominal and healthy.
* **Type II Error ($\beta$):** A **Missed Detection**. This occurs when a structural shift or localized crack has compromised the asset (rejects $H_1$), but the diagnostic system fails to alert the team (fails to reject $H_0$).

**Consequence of an Ultra-Conservative Threshold ($\alpha = 0.0001$)**

When $\alpha$ is reduced dramatically to eliminate false alarms, the critical value boundary of the test statistic is pushed deep into the extreme tails of the baseline distribution.

1. **Impact on Statistical Power ($1-\beta$):** Because the threshold for detecting anomalies becomes incredibly stringent, minor or emerging structural shifts will fail to cross the new partition line. Therefore, $\beta$ (missed detections) will sharply escalate, causing the **statistical power ($1-\beta$) of the model to collapse**.
2. **Impact on Confidence Ellipsoids:** The confidence level of the multi-dimensional boundary scales as $1-\alpha = 0.9999$. This means the **geometric volume of the healthy operational ellipsoid expands significantly**. The boundary swells to encompass almost all variation, making the system highly insensitive to genuine baseline degradation.


### Part B: Theoretical Extension


Let $\{X_n\}$ and $\{Y_n\}$ be sequences of random variables (or vectors). If $X_n$ converges in distribution to a random variable $X$ ($X_n \xrightarrow{d} X$), and $Y_n$ converges in probability to a constant scalar or matrix $c$ ($Y_n \xrightarrow{p} c$), then:

* $X_n + Y_n \xrightarrow{d} X + c$
* $X_n Y_n \xrightarrow{d} cX$
* $X_n / Y_n \xrightarrow{d} X/c$ (provided $c \neq 0$)

**Operational Distribution Justification**

The Lindeberg-Lévy Multivariate CLT asserts that:


$$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma})$$

Multiplying by the inverse matrix square root of the population covariance normalizes the space:


$$\boldsymbol{\Sigma}^{-1/2} \left[ \sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \right] \xrightarrow{d} \mathscr{N}(\mathbf{0}, \mathbf{I})$$

By the Weak Law of Large Numbers (WLLN), the unbiased empirical sample covariance matrix $\mathbf{S}$ is a continuous, consistent estimator of the population matrix $\boldsymbol{\Sigma}$. Therefore, as $n \to \infty$, $\mathbf{S}$ converges in probability to the true constant profile:


$$\mathbf{S} \xrightarrow{p} \boldsymbol{\Sigma} \implies \mathbf{S}^{-1/2} \xrightarrow{p} \boldsymbol{\Sigma}^{-1/2}$$

Applying **Slutsky's Theorem**, we multiply the sequence converging in distribution by the sequence converging in probability to a constant matrix:


$$\mathbf{S}^{-1/2} \left[ \sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \right] \xrightarrow{d} \boldsymbol{\Sigma}^{-1/2} \cdot \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma}) = \mathscr{N}(\mathbf{0}, \mathbf{I})$$

Because the target boundary converges asymptotically to a standard Gaussian space, we scale the localized coordinate parameters back to the empirical matrix metric:


$$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \sim \mathscr{N}(\mathbf{0}, \mathbf{S}) \implies \widehat{\boldsymbol{\mu}}_n \sim \mathscr{N}\left(\boldsymbol{\mu}, \frac{1}{n}\mathbf{S}\right)$$

This mathematically validates substituting $\mathbf{S}$ for $\boldsymbol{\Sigma}$ without corrupting the stability of the asymptotic diagnostic envelope.

 ###
 Part C: Numerical Verification

In [ ]:
import numpy as np
import pandas as pd
import scipy.stats as stats

# --- DO NOT MODIFY: Synthetic Strain Sensor Data Generator ---
np.random.seed(42)
n_samples = 5000
n_features = 3

base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[[0.09, 0.02, 0.01],
         [0.02, 0.06, 0.03],
         [0.01, 0.03, 0.05]],
    size=n_samples
)

base_data[4000:, 0] += 0.015
base_data[4000:, 2] -= 0.010

df_strain = pd.DataFrame(
    base_data,
    columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3']
)
# ----------------------------------------------------------------


def verify_first_moment_homogeneity(df: pd.DataFrame, g_chunks: int = 5) -> dict:
    """
    Partition data into consecutive chunks and test
    first-moment homogeneity using Wilks' Lambda.
    """

    X = df.values
    n, m = X.shape

    if n % g_chunks != 0:
        raise ValueError("Dataset size must be divisible by g_chunks")

    n_j = n // g_chunks

    # Global mean
    grand_mean = np.mean(X, axis=0)

    # Initialize matrices
    W = np.zeros((m, m))
    B = np.zeros((m, m))

    # Process each chunk
    for j in range(g_chunks):

        start = j * n_j
        end = (j + 1) * n_j

        chunk = X[start:end]
        chunk_mean = np.mean(chunk, axis=0)

        # Within-chunk SSCP
        centered = chunk - chunk_mean
        W += centered.T @ centered

        # Between-chunk SSCP
        mean_diff = (chunk_mean - grand_mean).reshape(m, 1)
        B += n_j * (mean_diff @ mean_diff.T)

    # Wilks' Lambda
    Lambda = np.linalg.det(W) / np.linalg.det(W + B)

    # Bartlett approximation
    chi2_calc = -(n - 1 - (m + g_chunks) / 2) * np.log(Lambda)

    # Degrees of freedom
    df_chi2 = m * (g_chunks - 1)

    # p-value
    p_value = 1 - stats.chi2.cdf(chi2_calc, df=df_chi2)

    conclusion = (
        "Reject H0: baseline center shifted over time."
        if p_value < 0.05
        else "Fail to reject H0: first moment is homogeneous."
    )

    return {
        "Wilks_Lambda": Lambda,
        "Bartlett_Chi2": chi2_calc,
        "df": df_chi2,
        "p_value": p_value,
        "conclusion": conclusion
    }


# Execute
results = verify_first_moment_homogeneity(df_strain, g_chunks=5)

for k, v in results.items():
    print(f"{k}: {v}")

Wilks_Lambda: 0.9904519796929591
Bartlett_Chi2: 47.9215049961488
df: 12
p_value: 3.2255259508895406e-06
conclusion: Reject H0: baseline center shifted over time.


# Results and Discussion

The dataset was partitioned into 5 consecutive, non-overlapping blocks of 1000 observations each. A MANOVA-based test for first-moment homogeneity was performed by computing the Within-Chunk matrix (W) and Between-Chunk matrix (B), followed by the evaluation of Wilks' Lambda.

### Calculated Statistics

Wilks' Lambda (Λ) = 0.9904519796929591

Bartlett Chi-Square Statistic = 47.9215049961488

Degrees of Freedom = 12

p-value = 3.2255259508895406 × 10⁻⁶

### Hypothesis Test

H₀: The joint mean vector remains constant across all 5 chunks (first-moment homogeneous).

H₁: At least one chunk has a different joint mean vector.

Using a significance level of α = 0.05:

p-value = 3.2255259508895406 × 10⁻⁶ < 0.05

Therefore, the null hypothesis is rejected.

### Conclusion

The MANOVA test indicates a statistically significant difference in the mean vectors among the five data blocks. Consequently, the assumption of Constant Joint Mean (First Moment Homogeneity) is not satisfied.

The baseline center has shifted over time, suggesting that the dataset exhibits temporal drift and does not fully satisfy the weak i.i.d. assumption required for a global PCA or Factor Analysis model.

# Geometric Subspace Optimization via Principal Component Analysis (PCA)

## Part A



To exploit the structure of the empirical covariance matrix for structural health monitoring, we map the raw random variables into an orthogonal, uncorrelated coordinate space. This is achieved via Principal Component Analysis (PCA) derived directly from the population moments.

Consider a zero-mean random varaible $\widetilde{\mathbf{X}}_i = (\mathbf{X}_i - \boldsymbol{\mu}) \in \mathbb{R}^{m}$, representing the centered baseline features. Its population covariance matrix $\boldsymbol{\Sigma} \in \mathbb{R}^{m \times m}$ is symmetric and positive-semidefinite:

$$\boldsymbol{\Sigma} = \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T]$$

By the Spectral Theorem (or the Singular Value Decomposition of a symmetric matrix), $\boldsymbol{\Sigma}$ can be factored into its orthogonal eigenvectors and real, non-negative eigenvalues:

$$\boldsymbol{\Sigma} = \mathbf{P} \mathbf{\Lambda} \mathbf{P}^T$$

Where:

* $\mathbf{P} = [\mathbf{p}_1, \mathbf{p}_2, \dots, \mathbf{p}_m] \in \mathbb{R}^{m \times m}$ is an orthogonal matrix ($\mathbf{P}^T\mathbf{P} = \mathbf{P}\mathbf{P}^T = \mathbf{I}$) containing the principal component loading vectors.
* $\mathbf{\Lambda} = \text{diag}(\lambda_1, \lambda_2, \dots, \lambda_m) \in \mathbb{R}^{m \times m}$ is a diagonal matrix containing sorted eigenvalues ($\lambda_1 \ge \lambda_2 \ge \dots \ge \lambda_m \ge 0$), representing the population variance along each principal axis.

---

We define the transformed random varable $\mathbf{Z}_i \in \mathbb{R}^{m \times 1}$ as the projection of the centered observations onto the orthogonal eigenvector basis:

$$\mathbf{Z}_i \triangleq \mathbf{P}^T \widetilde{\mathbf{X}}_i$$

To prove that this transformation completely de-correlates the feature space, we evaluate the theoretical covariance of $\mathbf{Z}_i$:

$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbb{E}[(\mathbf{P}^T \widetilde{\mathbf{X}}_i)(\mathbf{P}^T \widetilde{\mathbf{X}}_i)^T]$$

Utilizing the linear properties of expectation and the matrix transpose identity $(\mathbf{A}\mathbf{B})^T = \mathbf{B}^T\mathbf{A}^T$:

$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbf{P}^T \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T] \mathbf{P} = \mathbf{P}^T \boldsymbol{\Sigma} \mathbf{P}$$

Substituting the spectral decomposition $\boldsymbol{\Sigma} = \mathbf{P}\mathbf{\Lambda}\mathbf{P}^T$ into the equation yields the definitive diagonalization proof:

$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbf{P}^T (\mathbf{P} \mathbf{\Lambda} \mathbf{P}^T) \mathbf{P} = (\mathbf{P}^T \mathbf{P}) \mathbf{\Lambda} (\mathbf{P}^T \mathbf{P}) = \mathbf{I} \mathbf{\Lambda} \mathbf{I} = \mathbf{\Lambda}$$

Because $\mathbf{\Lambda}$ is strictly diagonal, the cross-covariances between distinct principal components are exactly zero ($\mathbb{E}[Z_{i,j}Z_{i,k}] = 0$ for all $j \neq k$).

---

In practical applications, the latent matrix $\boldsymbol{\Sigma}$ is unknown. We operationalize PCA by substituting the consistent, unbiased empirical sample covariance matrix $\mathbf{S}$ derived via our Maximum Likelihood framework:

$$\mathbf{S} = \widehat{\mathbf{P}} \widehat{\mathbf{\Lambda}} \widehat{\mathbf{P}}^T$$

The observed realizations are centered using the empirical mean vector $\widehat{\boldsymbol{\mu}}_n$ and projected into the empirical principal space to obtain sample scores $\mathbf{z}_i$:

$$\mathbf{z}_i = \widehat{\mathbf{P}}^T (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)$$

This operational mapping guarantees that the sample covariance of the transformed data matrix collapses to the diagonal matrix $\widehat{\mathbf{\Lambda}}$, establishing an optimal coordinate system for isolating independent physical fault modes in your diagnostics layer.

---

To compress the feature space from $m$ dimensions to a lower dimension $k$ (where $k < m$), we partition the empirical orthogonal basis matrix $\widehat{\mathbf{P}}$ into two orthogonal subspaces:$$\widehat{\mathbf{P}} = \begin{bmatrix} \widehat{\mathbf{P}}_k & \Big| & \widehat{\mathbf{P}}_{m-k} \end{bmatrix}$$where $\widehat{\mathbf{P}}_k \in \mathbb{R}^{m \times k}$ contains the first $k$ column eigenvectors corresponding to the largest $k$ eigenvalues ($\widehat{\lambda}_1 \ge \dots \ge \widehat{\lambda}_k$). This partitioning allows us to partition $\mathbf{z}\in \mathbb{R}^m$ as
$$\mathbf{z}_{i}=\begin{bmatrix}\mathbf{z}_{i,k} \\\mathbf{z}_{i,m-k}
\end{bmatrix}$$
where $\mathbf{z}_{i,k} \in \mathbb{R}^{k}$ and $\mathbf{z}_{i,m-k} \in \mathbb{R}^{m-k}.$
From
$$\mathbf{z}_i = \widehat{\mathbf{P}}^T (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)$$
we thus have
$$(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)=\widehat{\mathbf{P}}_k \mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}.$$

The partitioning of the empirical basis matrix allows us to isolate the deterministic structural dynamics from the unmodeled residual noise. By retaining only the first $k$ principal components, we define the reconstructed feature vector $\widehat{\mathbf{x}}_i \in \mathbb{R}^m$ as the projection of the low-dimensional scores back into the original measurement space:

$$\widehat{\mathbf{x}}_i \triangleq \widehat{\boldsymbol{\mu}}_n + \widehat{\mathbf{P}}_k \mathbf{z}_{i,k}$$

The geometric displacement between the actual observation $\mathbf{x}_i$ and its lower-dimensional reconstruction $\widehat{\mathbf{x}}_i$ defines the residual error vector $\mathbf{e}_i \in \mathbb{R}^m$:

$$\mathbf{e}_i = \mathbf{x}_i - \widehat{\mathbf{x}}_i = \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}$$

Because $\widehat{\mathbf{P}}_k$ and $\widehat{\mathbf{P}}_{m-k}$ span mutually orthogonal subspaces, the total energy of the centered observation vector can be cleanly decomposed via the Pythagorean identity:

$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \|\widehat{\mathbf{P}}_k \mathbf{z}_{i,k}\|^2 + \|\widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}\|^2 = \|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2$$

---

To construct a robust diagnostics layer, we monitor changes in both the principal and residual subspaces. This is achieved using two complementary statistical indicators.

1. Hotelling's $T^2$ Statistic (Principal Subspace)

Hotelling’s $T^2$ statistic measures the Mahalanobis distance within the $k$-dimensional principal subspace. It detects significant shifts or anomalies in the nominal, high-variance operational modes of the structure:

$$T^2_i = \mathbf{z}_{i,k}^T \widehat{\mathbf{\Lambda}}_k^{-1} \mathbf{z}_{i,k} = \sum_{j=1}^k \frac{z_{i,j}^2}{\widehat{\lambda}_j}$$

where $\widehat{\mathbf{\Lambda}}_k = \text{diag}(\widehat{\lambda}_1, \dots, \widehat{\lambda}_k)$. Under the assumption of multivariate normality, a statistically significant inflation in $T^2_i$ indicates that the system has drifted away from its baseline operational equilibrium.

2. The $Q$ Statistic / Squared Prediction Error (Residual Subspace)

The $Q$ statistic (also known as the Squared Prediction Error, SPE) measures the magnitude of the residual vector. It quantifies how well the baseline $k$-dimensional model explains a new observation:

$$Q_i = \|\mathbf{e}_i\|^2 = \mathbf{e}_i^T \mathbf{e}_i = \mathbf{z}_{i,m-k}^T \mathbf{z}_{i,m-k} = \sum_{j=k+1}^m z_{i,j}^2$$

Anomalies flagged by the $Q$ statistic indicate a break in the underlying structural correlation structure. This implies that a new, unmodeled physical phenomenon—such as a localized structural crack or sensor degradation—is projecting energy into the residual noise floor.

---

To optimize the diagnostic layer for dimensionality reduction or noise filtering, we must formally quantify the information captured by a truncated subset of the principal components. Because the trace of a matrix is invariant under cyclic permutations, the total variance of the system is preserved when mapping from the raw feature space to the de-correlated empirical principal space:

$$\text{tr}(\mathbf{S}) = \text{tr}(\widehat{\mathbf{P}}\widehat{\mathbf{\Lambda}}\widehat{\mathbf{P}}^T) = \text{tr}(\widehat{\mathbf{\Lambda}}\widehat{\mathbf{P}}^T\widehat{\mathbf{P}}) = \text{tr}(\widehat{\mathbf{\Lambda}}) = \sum_{j=1}^m \widehat{\lambda}_j$$

By partitioning the basis matrix into $\widehat{\mathbf{P}}_k$ and $\widehat{\mathbf{P}}_{m-k}$ to isolate the first $k$ dominant components, we explicitly divide this total system variance into the principal (explained) and residual (unexplained) subspaces:

* **Marginal and Cumulative Explained Variance:** The proportion of total variance captured within the principal subspace spanned by $\widehat{\mathbf{P}}_k$ is governed by the largest $k$ eigenvalues. Accumulating these ratios yields the cumulative explained variance ratio $\Phi(k)$, which measures the informational content of the low-dimensional score vector $\mathbf{z}_{i,k}$:

$$\Phi(k) = \frac{\sum_{j=1}^k \widehat{\lambda}_j}{\sum_{i=1}^m \widehat{\lambda}_i}$$

* **Residual Unexplained Variance:** Conversely, the information discarded by omitting the remaining $m-k$ low-variance axes represents the residual unexplained variance ratio $\Psi(k)$. This metric explicitly gauges the structural scale of the unmodeled operational subspace spanned by $\widehat{\mathbf{P}}_{m-k}$, tracking the total energy remaining in the residual score vector $\mathbf{z}_{i,m-k}$:

$$\Psi(k) = 1.0 - \Phi(k) = \frac{\sum_{j=k+1}^m \widehat{\lambda}_j}{\sum_{i=1}^m \widehat{\lambda}_i}$$

Operationalizing these metrics using the empirical eigenvalues $\widehat{\lambda}_j$ provides a mathematically rigorous thresholding strategy for our subspace decomposition. By plotting the decay of $\widehat{\lambda}_j$ against $\Psi(k)$, you can isolate the precise point $k$ where the system's true deterministic mechanical behavior—reconstructed via $\widehat{\mathbf{P}}_k \mathbf{z}_{i,k}$—transitions into the unmodeled stochastic white noise absorbed by $\widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}$.

## Part B

In [ ]:
import numpy as np
import pandas as pd


class DiagnosticsAnalyzer:

    def analyze_t2_q_statistics(self, X):
        """
        Evaluate Hotelling's T² and Q statistics for k = 1, 2, 3.

        Parameters
        ----------
        X : numpy.ndarray or pandas.DataFrame
            Data matrix of shape (n_samples, n_features)

        Returns
        -------
        dict
            Expected values of T² and Q for each k.
        """

        # Convert to numpy array
        X = np.asarray(X)

        n, m = X.shape

        # --------------------------------------------------
        # 1. Standardize and center data
        # --------------------------------------------------
        mean_vec = np.mean(X, axis=0)
        std_vec = np.std(X, axis=0, ddof=1)

        X_std = (X - mean_vec) / std_vec

        # --------------------------------------------------
        # 2. Unbiased covariance matrix
        # --------------------------------------------------
        S = np.cov(X_std, rowvar=False, ddof=1)

        # --------------------------------------------------
        # 3. Eigen decomposition
        # --------------------------------------------------
        eigvals, eigvecs = np.linalg.eigh(S)

        # Sort descending
        idx = np.argsort(eigvals)[::-1]

        eigvals = eigvals[idx]
        eigvecs = eigvecs[:, idx]

        # --------------------------------------------------
        # 4. PCA scores
        # --------------------------------------------------
        Z = X_std @ eigvecs

        results = {}

        # --------------------------------------------------
        # 5. Loop through k = 1, 2, 3
        # --------------------------------------------------
        for k in [1, 2, 3]:

            # Hotelling T²
            T2 = np.sum(
                (Z[:, :k] ** 2) / eigvals[:k],
                axis=1
            )

            # Residual Q statistic
            Q = np.sum(
                Z[:, k:] ** 2,
                axis=1
            )

            results[f"k={k}"] = {
                "Mean_T2": np.mean(T2),
                "Mean_Q": np.mean(Q)
            }

        return results

#  Latent Subspace Decomposition via Factor Analysis (FA)



# Part A: Theoretical Fundamentals
Let $\mathbf{Z}_i \in \mathbb{R}^m$ be the normalized, zero-mean random variable derived from centering and scaling the $i$-th baseline feature vector $\mathbf{X}_i$:

$$\mathbf{Z}_i = \mathbf{D}^{-1/2}(\mathbf{X}_i - \boldsymbol{\mu})$$

where $\boldsymbol{\mu}$ is the universal population mean and $\mathbf{D} = \text{diag}(\sigma_1^2, \sigma_2^2, \dots, \sigma_m^2)$ is the diagonal matrix containing the population variances of $\mathbf{X}_i$. By construction, the theoretical population covariance matrix of $\mathbf{Z}_i$ collapses to the universal population correlation matrix, $\mathbf{R}$:

$$\mathbf{R} = \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] \in \mathbb{R}^{m \times m}, \quad \text{with} \quad \mathbb{E}[Z_{i,j}^2] = 1.0 \quad \forall \ j \in \{1, \dots, m\}.$$

We assume that the $m$ visible features in $\mathbf{Z}_i$ are linearly generated by a heavily compressed subset of $k$ unobservable (latent) physical mechanisms or operational conditions, where $k < m$. Formally, the generative model for the $i$-th realization vector is defined as:

$$\mathbf{Z}_i = \boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i$$

Where:

* $\mathbf{F}_i = [F_{i,1}, F_{i,2}, \dots, F_{i,k}]^T \in \mathbb{R}^k$ is the random variable containing the **common factors** that govern the global physical state of the structure.
* $\boldsymbol{\Lambda} \in \mathbb{R}^{m \times k}$ is the invariant **factor loading matrix**, where the scalar $\lambda_{j,r}$ quantifies the linear sensitivity of the $j$-th feature to the $r$-th latent structural mode.
* $\boldsymbol{\epsilon}_i = [\epsilon_{i,1}, \epsilon_{i,2}, \dots, \epsilon_{i,m}]^T \in \mathbb{R}^m$ is the random variable representing the **specific errors** or **uniquenesses** unique to each measurement channel.

To ensure mathematical tractability and structurally isolate global physical effects from local disturbances, we impose three orthogonal population constraints:

1. **Orthogonal Common Factors:** The latent structural modes are mutually de-correlated and normalized to unit variance:

$$\mathbb{E}[\mathbf{F}_i] = \mathbf{0}, \quad \mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] = \mathbf{I}_{k \times k}$$


2. **De-correlated Unique Errors:** The measurement noise or local variance of any given feature is statistically independent of all other features. Consequently, the unique error covariance matrix $\boldsymbol{\Psi}$ must be strictly diagonal:

$$\mathbb{E}[\boldsymbol{\epsilon}_i] = \mathbf{0}, \quad \mathbb{E}[\boldsymbol{\epsilon}_i \boldsymbol{\epsilon}_i^T] = \boldsymbol{\Psi} = \text{diag}(\varphi_1^2, \varphi_2^2, \dots, \varphi_m^2)$$


3. **Subspace Orthogonality:** The global physical mechanics driving the common factors exert zero statistical influence over the local, channel-specific unique errors:

$$\mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T] = \mathbf{0}_{m \times k}.$$

Under the weight of these constraints, the theoretical correlation structure of the system can be decomposed directly. Expanding the expectation of the product $\mathbf{Z}_i\mathbf{F}_i^T$ yields:

$$\mathbb{E}[\mathbf{Z}_i \mathbf{F}_i^T] = \mathbb{E}[(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)\mathbf{F}_i^T] = \boldsymbol{\Lambda}\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T] + \mathbb{E}[\boldsymbol{\epsilon}_i\mathbf{F}_i^T] = \boldsymbol{\Lambda}\mathbf{I} + \mathbf{0} = \boldsymbol{\Lambda}$$

Thus, the factor loading matrix $\boldsymbol{\Lambda}$ explicitly maps the cross-correlation between the observable variables and the hidden structural modes. We evaluate the total population correlation matrix $\mathbf{R}$ by expanding the generative formula:

$$\mathbf{R} = \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbb{E}[(\boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i)(\boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i)^T]$$

$$\mathbf{R} = \boldsymbol{\Lambda} \mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] \boldsymbol{\Lambda}^T + \boldsymbol{\Lambda}\mathbb{E}[\mathbf{F}_i \boldsymbol{\epsilon}_i^T] + \mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T]\boldsymbol{\Lambda}^T + \mathbb{E}[\boldsymbol{\epsilon}_i \boldsymbol{\epsilon}_i^T]$$

Utilizing the orthogonality properties ($\mathbf{I}$, $\mathbf{0}$, and $\boldsymbol{\Psi}$), the expression cleanly reduces to the **Fundamental Equation of Factor Analysis**:

$$\mathbf{R} = \boldsymbol{\Lambda} \boldsymbol{\Lambda}^T + \boldsymbol{\Psi}.$$

Thus factor analysis reduces to a marix factorization problem.

**Estimating the Loading Matrix $\boldsymbol{\Lambda}$ via Maximum Likelihood Optimization:**

In the following we establish how this loading matrix is determined from the empirical data. In a standard framework, the parameters are estimated by maximizing the joint log-likelihood of the observed data, assuming a multivariate normal distribution.

Given our underlying generative model where $\mathbf{Z}_i \sim \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma})$ with the structured covariance $\boldsymbol{\Sigma} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$, the objective function is derived by minimizing the structural discrepancy between the fitted model covariance ($\boldsymbol{\Sigma}$) and the empirical correlation matrix ($\mathbf{R}$).

The log-likelihood objective function to be minimized with respect to $\boldsymbol{\Lambda}$ and $\boldsymbol{\Psi}$ is formulated as:

$$\mathscr{L}(\boldsymbol{\Lambda}, \boldsymbol{\Psi}) = \ln|\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}| + \text{Tr}\left[ \mathbf{R} (\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi})^{-1} \right] - \ln|\mathbf{R}| - m$$

To isolate $\boldsymbol{\Lambda}$ during an optimization step, we take the partial derivative of the log-likelihood function with respect to $\boldsymbol{\Lambda}$ and equate it to zero. Under a fixed uniqueness field $\boldsymbol{\Psi}$, this yields the algebraic system:

$$\frac{\partial \mathscr{L}}{\partial \boldsymbol{\Lambda}} = 2(\boldsymbol{\Sigma}^{-1} - \boldsymbol{\Sigma}^{-1}\mathbf{R}\boldsymbol{\Sigma}^{-1})\boldsymbol{\Lambda} = \mathbf{0}$$

$$\boldsymbol{\Sigma}^{-1}\boldsymbol{\Lambda} = \boldsymbol{\Sigma}^{-1}\mathbf{R}\boldsymbol{\Sigma}^{-1}\boldsymbol{\Lambda}$$

Multiplying both sides by $(\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi})$, we decouple the expression to project $\mathbf{R}$ directly into the scaled parameter space:

$$\mathbf{R}\boldsymbol{\Sigma}^{-1}\boldsymbol{\Lambda} = \boldsymbol{\Lambda}$$

$$\mathbf{R}(\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi})^{-1}\boldsymbol{\Lambda} = \boldsymbol{\Lambda}$$

Using the Woodbury matrix identity to simplify the inverse of the sum, the conditional maximization step yields the exact structural transformation:

$$\boldsymbol{\Lambda} = \boldsymbol{\Psi}^{1/2} \mathbf{P}_k (\boldsymbol{\Lambda}_k - \mathbf{I})^{1/2}$$

where $\mathbf{P}_k \in \mathbb{R}^{m \times k}$ and $\boldsymbol{\Lambda}_k \in \mathbb{R}^{k \times k}$ represent the top $k$ eigenvectors and diagonalized eigenvalues extracted directly from the uniqueness-scaled empirical correlation matrix:

$$\boldsymbol{\Psi}^{-1/2} \mathbf{R} \boldsymbol{\Psi}^{-1/2} = \mathbf{P} \boldsymbol{\Lambda} \mathbf{P}^T$$

Because $\boldsymbol{\Psi}$ is initially unknown, the engine applies an iterative Expectation-Maximization (EM) or conditional coordinate descent loop. It alternates between computing the loading matrix $\boldsymbol{\Lambda}$ via this spectral solution and updating the diagonal noise floor matrix via $\boldsymbol{\Psi} = \text{diag}(\mathbf{R} - \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T)$ until numerical convergence is achieved.

# 3)



1. Joint Subspace Covariance Construction

To find the conditional expectation $\mathbb{E}[\mathbf{F}_i | \mathbf{z}_i]$, we first examine how the observable normalized variables $\mathbf{Z}_i$ and the hidden common factors $\mathbf{F}_i$ vary together. We construct a partitioned joint random vector $\mathbf{Y}_i$:

$$\mathbf{Y}_i = \begin{bmatrix} \mathbf{F}_i \\ \mathbf{Z}_i \end{bmatrix}$$

Because both $\mathbf{Z}_i$ and $\mathbf{F}_i$ have a mean of zero ($\mathbb{E}[\mathbf{Z}_i] = \mathbf{0}$ and $\mathbb{E}[\mathbf{F}_i] = \mathbf{0}$), the joint mean is $\mathbb{E}[\mathbf{Y}_i] = \mathbf{0}$.

Now, let's construct the partitioned joint covariance matrix $\boldsymbol{\Sigma}_{YY}$:

$$\boldsymbol{\Sigma}_{YY} = \mathbb{E}[\mathbf{Y}_i \mathbf{Y}_i^T] = \begin{bmatrix} \mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] & \mathbb{E}[\mathbf{F}_i \mathbf{Z}_i^T] \\ \mathbb{E}[\mathbf{Z}_i \mathbf{F}_i^T] & \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] \end{bmatrix}$$

From the foundations of Factor Analysis that was established before, we know the values of these individual blocks:

1. $\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbf{R}$ (The visible correlation matrix)
2. $\mathbb{E}[\mathbf{Z}_i \mathbf{F}_i^T] = \boldsymbol{\Lambda}$ (The factor loading matrix, which we proved earlier)
3. $\mathbb{E}[\mathbf{F}_i \mathbf{Z}_i^T] = (\mathbb{E}[\mathbf{Z}_i \mathbf{F}_i^T])^T = \boldsymbol{\Lambda}^T$
4. $\mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] = \mathbf{I}_{k \times k}$ (The orthogonal common factor constraint)

Substituting these blocks into our partitioned matrix yields:

$$\boldsymbol{\Sigma}_{YY} = \begin{bmatrix} \mathbf{I} & \boldsymbol{\Lambda}^T \\ \boldsymbol{\Lambda} & \mathbf{R} \end{bmatrix}$$

---

2. Applying the Conditional Expectation Theorem

Under the standard assumption of a Multivariate Normal distribution for the statistical inference layer, the conditional distribution of one partitioned segment given the other is well-defined.

For any joint normal system where $\boldsymbol{\Sigma} = \begin{bmatrix} \boldsymbol{\Sigma}_{11} & \boldsymbol{\Sigma}_{12} \\ \boldsymbol{\Sigma}_{21} & \boldsymbol{\Sigma}_{22} \end{bmatrix}$, the conditional expectation of the second slot given a realization of the first slot ($\mathbf{x}_1$) is given by the classical linear regression projection formula:

$$\mathbb{E}[\mathbf{X}_1 | \mathbf{X}_2] = \boldsymbol{\mu}_1 + \boldsymbol{\Sigma}_{12} \boldsymbol{\Sigma}_{22}^{-1} (\mathbf{X}_2 - \boldsymbol{\mu}_2)$$

Mapping this directly to our factor analysis matrix partitions:

* Slot 1 is our hidden latent factor: $\mathbf{X}_1 = \mathbf{F}_i$ (with $\boldsymbol{\mu}_1 = \mathbf{0}$)
* Slot 2 is our observed sample realization: $\mathbf{X}_2 = \mathbf{Z}_i$ (with $\boldsymbol{\mu}_2 = \mathbf{0}$)
* The cross-covariance section $\boldsymbol{\Sigma}_{12} = \boldsymbol{\Lambda}^T$
* The observed covariance section $\boldsymbol{\Sigma}_{22} = \mathbf{R}$

Substituting these variables into the projection formula yields the factor score realization $\mathbf{f}_i$:

$$\mathbf{f}_i = \mathbb{E}[\mathbf{F}_i | \mathbf{Z}_i] = \mathbf{0} + \boldsymbol{\Lambda}^T \mathbf{R}^{-1} (\mathbf{Z}_i - \mathbf{0})$$

$$\mathbf{f}_i = \boldsymbol{\Lambda}^T \mathbf{R}^{-1} \mathbf{Z}_i$$


Because this estimator is derived via the conditional expectation, it possesses a vital property: it is the **Best Linear Unbiased Predictor (BLUP)**.

Even if the data deviates slightly from strict normality, this specific linear matrix transformation minimizes the mean squared error ($\mathbb{E}[\|\mathbf{F}_i - \mathbf{f}_i\|^2]$) between the hidden, true physical state of the structure and the feature estimates.

#  Subspace Diagnostics & Feature De-correlation

In [4]:
import numpy as np
import pandas as pd
from typing import Optional, Sequence, Dict, Any
from sklearn.decomposition import FactorAnalysis
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# =====================================================================
# COMBINED SUBSPACE ENGINES (PCA & FA PIPELINES)
# =====================================================================

class DualSubspaceDiagnosticsEngine:
    def __init__(self, df: pd.DataFrame):
        self.df = df

    def compute_empirical_pca(self, columns: Optional[Sequence[str]] = None, show_plot: bool = True) -> Dict[str, Any]:
        """
        Operationalizes the geometric de-correlation framework via PCA.
        Decomposes the unbiased empirical covariance matrix S into its orthogonal
        basis P_hat and renders a clean 2x3 optimization and feature contribution dashboard.
        """
        if self.df is None:
            raise ValueError("Error: No data loaded.")

        if columns is None:
            target_cols = self.df.select_dtypes(include=[np.number]).columns.tolist()
            if 'count' in target_cols: target_cols.remove('count')
            target_cols = list(target_cols)
        else:
            target_cols = list(columns)

        X = self.df[target_cols].copy().dropna().values
        n, m = X.shape

        if n <= m:
            raise ValueError("Sample size n must be larger than feature dimensions m to isolate principal components.")

        # 1. Centering
        mu_hat = np.mean(X, axis=0)
        X_centered = X - mu_hat

        # 2. Compute Unbiased Sample Covariance Matrix S
        S_matrix = np.cov(X, rowvar=False, ddof=1)

        # 3. Spectral Decomposition
        eigenvalues, eigenvectors = np.linalg.eigh(S_matrix)
        idx = np.argsort(eigenvalues)[::-1]
        lambda_hat = eigenvalues[idx]
        P_hat = eigenvectors[:, idx]
        lambda_hat = np.clip(lambda_hat, a_min=1e-15, a_max=None)

        # 4. Variance Information Calculations
        total_variance = np.sum(lambda_hat)
        if total_variance == 0:
            raise ValueError("Total variance is zero.")

        explained_variance_ratio = lambda_hat / total_variance
        cumulative_variance_ratio = np.cumsum(explained_variance_ratio)
        unexplained_variance_ratio = 1.0 - cumulative_variance_ratio

        # 5. Map Realizations to Principal Component Scores
        Z_scores = np.dot(X_centered, P_hat)
        S_Z = np.cov(Z_scores, rowvar=False, ddof=1)

        # 6. Evaluate Statistical Distance Metrics (T^2 and Q)
        k_range = np.arange(1, m)
        mean_T2_vs_k = []
        mean_Q_vs_k = []
        T2_matrix = np.zeros((n, len(k_range)))
        Q_matrix = np.zeros((n, len(k_range)))

        for idx_k, k in enumerate(k_range):
            Z_k = Z_scores[:, :k]
            lambda_k = lambda_hat[:k]
            T2_samples = np.sum((Z_k ** 2) / lambda_k, axis=1)
            T2_matrix[:, idx_k] = T2_samples
            mean_T2_vs_k.append(np.mean(T2_samples))

            Z_residual = Z_scores[:, k:]
            Q_samples = np.sum(Z_residual ** 2, axis=1)
            Q_matrix[:, idx_k] = Q_samples
            mean_Q_vs_k.append(np.mean(Q_samples))

        print(f"\n--- Operationalizing Geometric De-correlation Layer (PCA) ---")
        print(f"Decomposing structural space of {m} features using {n} samples.")
        print(f"Total System Variance (Trace[S]): {total_variance:.4f}")

        # 7. Enhanced 2x3 Diagnostic Dashboard
        if show_plot:
            pc_labels = [f"PC {i+1}" for i in range(m)]
            k_labels = [f"k={k}" for k in k_range]
            sensor_labels = target_cols

            fig = make_subplots(
                rows=2, cols=3,
                horizontal_spacing=0.18,
                vertical_spacing=0.28,
                subplot_titles=(
                    "Feature Loading Matrix |P_hat|",
                    "Component Values (Eigenvalues λ)",
                    "Information Profile (Explained Var.)",
                    "Residual Space (Unexplained Var.)",
                    "Mean Hotelling's T² vs Subspace Size k",
                    "Mean Q Statistic (SPE) vs Subspace Size k"
                )
            )

            # --- ROW 1, COL 1: PCA Loading Heatmap (Left-anchored Colorbar) ---
            fig.add_trace(
                go.Heatmap(
                    z=np.abs(P_hat),
                    x=pc_labels,
                    y=sensor_labels,
                    colorscale='YlOrRd',
                    colorbar=dict(
                        title="Loading Weight",
                        x=-0.12,          # Perfectly isolates colorbar to the left side
                        len=0.38,         # Fits height bounds of row 1
                        y=0.78,           # Centers vertically inside row 1
                        yanchor="middle",
                        xanchor="right",
                        titleside="top"
                    ),
                    showscale=True,
                    showlegend=False
                ),
                row=1, col=1
            )
            fig.update_xaxes(title_text="Principal Axes", row=1, col=1)

            # --- ROW 1, COL 2: Absolute Eigenvalues ---
            fig.add_trace(go.Bar(x=pc_labels, y=lambda_hat, name="Eigenvalue (λ_j)", marker=dict(color='#1f77b4', line=dict(color='black', width=0.5)), legendgroup="eigen"), row=1, col=2)
            fig.update_yaxes(title_text="Variance Magnitude", row=1, col=2)
            fig.update_xaxes(title_text="Principal Axes", row=1, col=2)

            # --- ROW 1, COL 3: Variance Profiles ---
            fig.add_trace(go.Bar(x=pc_labels, y=explained_variance_ratio * 100, name="Marginal Explained", marker=dict(color='#ff7f0e', opacity=0.75), legendgroup="expl"), row=1, col=3)
            fig.add_trace(go.Scatter(x=pc_labels, y=cumulative_variance_ratio * 100, mode='lines+markers', name='Cumulative Captured', line=dict(color='#d62728', width=2.5, dash='dash'), legendgroup="expl"), row=1, col=3)
            fig.update_yaxes(title_text="Captured Structure (%)", range=[-2, 105], row=1, col=3)
            fig.update_xaxes(title_text="Principal Axes", row=1, col=3)

            # --- ROW 2, COL 1: Remaining Noise ---
            fig.add_trace(go.Bar(x=pc_labels, y=unexplained_variance_ratio * 100, name="Remaining Noise", marker=dict(color='#2ca02c', line=dict(color='black', width=0.5)), legendgroup="noise"), row=2, col=1)
            fig.update_yaxes(title_text="Excluded Info (%)", range=[-2, 105], row=2, col=1)
            fig.update_xaxes(title_text="Principal Axes", row=2, col=1)

            # --- ROW 2, COL 2: Hotelling's T^2 ---
            fig.add_trace(go.Scatter(x=k_labels, y=mean_T2_vs_k, mode='lines+markers', name='Mean T²', line=dict(color='#9467bd', width=2.5), marker=dict(size=6, symbol='diamond'), legendgroup="t2"), row=2, col=2)
            fig.update_yaxes(title_text="Average T² Metric", row=2, col=2)
            fig.update_xaxes(title_text="Truncation Cutoff (k)", row=2, col=2)

            # --- ROW 2, COL 3: SPE / Q Statistic ---
            fig.add_trace(go.Scatter(x=k_labels, y=mean_Q_vs_k, mode='lines+markers', name='Mean Q (SPE)', line=dict(color='#e377c2', width=2.5), marker=dict(size=6, symbol='square'), legendgroup="q_stat"), row=2, col=3)
            fig.update_yaxes(title_text="Average Residual Energy", row=2, col=3)
            fig.update_xaxes(title_text="Truncation Cutoff (k)", row=2, col=3)

            # Global Unified Layout Configurations
            fig.update_layout(
                title=dict(text="Principal Component Analysis (PCA) Optimization & Feature Loading Dashboard", x=0.5, y=0.97, xanchor="center", yanchor="top"),
                template="plotly_white",
                showlegend=True,
                legend=dict(
                    orientation="h",
                    yanchor="bottom",
                    y=1.02,
                    xanchor="center",
                    x=0.5
                ),
                margin=dict(t=150, b=60, l=140, r=80),
                height=750,
                width=1250
            )
            fig.show()

        return {
            "mean_vector": mu_hat, "covariance_matrix_S": S_matrix, "eigenvalues_lambda": lambda_hat,
            "eigenvectors_P": P_hat, "explained_variance_ratio": explained_variance_ratio,
            "cumulative_variance_ratio": cumulative_variance_ratio, "unexplained_variance_ratio": unexplained_variance_ratio,
            "transformed_scores_Z": Z_scores, "score_covariance_diagonal": np.diag(S_Z),
            "features": target_cols, "k_values": k_range, "T2_matrix_vs_k": T2_matrix, "Q_matrix_vs_k": Q_matrix,
            "mean_T2_profile": np.array(mean_T2_vs_k), "mean_Q_profile": np.array(mean_Q_vs_k)
        }

    def compute_empirical_fa(self, k: int, columns: Optional[Sequence[str]] = None, show_plot: bool = True) -> Dict[str, Any]:
        """
        Operationalizes the Factor Analysis latent subspace framework with a clean 2x2 grid.
        Anchors the heatmap colorbar perfectly to the right edge of the first subplot.
        """
        if self.df is None:
            raise ValueError("Error: No data loaded.")

        if columns is None:
            target_cols = self.df.select_dtypes(include=[np.number]).columns.tolist()
            if 'count' in target_cols: target_cols.remove('count')
            target_cols = list(target_cols)
        else:
            target_cols = list(columns)

        X = self.df[target_cols].copy().dropna().values
        n, m = X.shape

        # 1. Standardize realizations to construct Z space
        mu_hat = np.mean(X, axis=0)
        std_hat = np.std(X, axis=0, ddof=1)
        std_hat[std_hat == 0] = 1e-15
        Z = (X - mu_hat) / std_hat
        R_matrix = np.corrcoef(X, rowvar=False)

        # 2. Fit Factor Analysis Model (Varimax Rotation)
        fa = FactorAnalysis(n_components=k, rotation='varimax', random_state=42)
        fa.fit(Z)
        lambda_matrix = fa.components_.T
        psi_diagonal = fa.noise_variance_

        # 3. Calculate Communality (h²) and Uniqueness (φ²) vectors
        communality = np.sum(lambda_matrix**2, axis=1)
        uniqueness = psi_diagonal

        # 4. Map realizations using Thomson's Regression Method
        F_scores = fa.transform(Z)

        print(f"\n--- Operationalizing Latent Subspace Layer (Factor Analysis) ---")
        print(f"Decomposing structural space of {m} sensors into {k} hidden physical factors.")
        print(f"Average System Communality (Shared Subspace Energy): {np.mean(communality)*100:.2f}%")
        print(f"Average System Uniqueness (Localized Noise Floor): {np.mean(uniqueness)*100:.2f}%")

        # 5. Clean 2x2 Dashboard
        if show_plot:
            sensor_labels = target_cols
            factor_labels = [f"Factor {j+1}" for j in range(k)]

            fig = make_subplots(
                rows=2, cols=2, horizontal_spacing=0.24, vertical_spacing=0.28,
                subplot_titles=(
                    "Structural Loadings Matrix |λ_(j,r)|",
                    "Variance Partitioning (Communality vs Uniqueness)",
                    "Sensor Uniqueness Noise Floor (φ²)",
                    "Latent Factor Scores Empirical Variance"
                )
            )

            # --- Subplot 1: Factor Loadings Heatmap (Colorbar repositioned to x=0.42) ---
            fig.add_trace(
                go.Heatmap(
                    z=np.abs(lambda_matrix),
                    x=factor_labels,
                    y=sensor_labels,
                    colorscale='YlOrRd',
                    colorbar=dict(
                        title="Sensitivity",
                        x=-0.1,          # Shifts colorbar out of the far right to center-left
                        len=0.38,         # Scales down length to match row 1 height bounds
                        y=0.78,          # Centers vertically inside row 1
                        yanchor="middle",
                        xanchor="right"
                    ),
                    showscale=True,
                    showlegend=False
                ),
                row=1, col=1
            )
            fig.update_xaxes(title_text="Latent Structures", row=1, col=1)

            # --- Subplot 2: Variance Partitioning ---
            fig.add_trace(go.Bar(y=sensor_labels, x=communality * 100, name="Communality (h² - Shared Structure)", orientation='h', marker=dict(color='#1f77b4'), legendgroup="communality"), row=1, col=2)
            fig.add_trace(go.Bar(y=sensor_labels, x=uniqueness * 100, name="Uniqueness (φ² - Channel Noise)", orientation='h', marker=dict(color='#ff7f0e'), legendgroup="uniqueness_bar"), row=1, col=2)
            fig.update_layout(barmode='stack')
            fig.update_xaxes(title_text="Variance Allocation (%)", range=[0, 100], row=1, col=2)

            # --- Subplot 3: Uniqueness Line Profile ---
            fig.add_trace(go.Scatter(x=sensor_labels, y=uniqueness, mode='lines+markers', name='Uniqueness Profile (φ²)', line=dict(color='#d62728', width=2, dash='dot'), marker=dict(size=8, symbol='x'), legendgroup="uniqueness_line"), row=2, col=1)
            fig.update_yaxes(range=[-0.05, 1.05], row=2, col=1)
            fig.update_xaxes(title_text="Monitored Channels", tickangle=25, row=2, col=1)

            # --- Subplot 4: Latent Factor Scores Empirical Variance ---
            factor_variances = np.var(F_scores, axis=0, ddof=1)
            fig.add_trace(go.Bar(x=factor_labels, y=factor_variances, name="Factor Empirical Variance", marker=dict(color='#2ca02c', line=dict(color='black', width=0.5)), legendgroup="variance"), row=2, col=2)
            fig.update_yaxes(title_text="Variance Level", row=2, col=2)
            fig.update_xaxes(title_text="Latent Vectors", row=2, col=2)

            # Global Layout Updates
            fig.update_layout(
                title=dict(text="Factor Analysis (FA) Latent Subspace Diagnostics Dashboard", x=0.5, y=0.97, xanchor="center", yanchor="top"),
                template="plotly_white",
                showlegend=True,
                legend=dict(
                    orientation="h",
                    yanchor="bottom",
                    y=1.02,
                    xanchor="center",
                    x=0.5,
                    traceorder="normal",
                    itemwidth=40,
                    itemsizing="constant"
                ),
                margin=dict(t=100, b=60, l=80, r=80),
                height=750,
                width=1100
            )
            fig.show()

        return {
            "mean_vector_mu": mu_hat, "std_vector_D": std_hat, "correlation_matrix_R": R_matrix,
            "factor_loadings_lambda": lambda_matrix, "uniqueness_psi": uniqueness,
            "communality_h2": communality, "latent_factor_scores_F": F_scores, "sensors": target_cols
        }

# =====================================================================
# DATA GENERATOR & PIPELINE EXECUTION
# =====================================================================
if __name__ == "__main__":
    # Set execution seed for perfect reproducible properties
    np.random.seed(42)
    n_samples = 2500

    # 1. Generate underlying, independent true physical driving forces (Latent Factors)
    f1 = np.random.normal(0, 1, n_samples)  # Primary structural mode/load
    f2 = np.random.normal(0, 1, n_samples)  # Secondary operational mode

    # 2. Construct the Observable Multi-Sensor Array with targeted noise profiles
    # Sensors 1 and 2 couple strongly to the primary process (f1)
    s1 = 0.85 * f1 + 0.10 * f2 + np.random.normal(0, 0.3, n_samples)
    s2 = 0.80 * f1 + 0.15 * f2 + np.random.normal(0, 0.35, n_samples)

    # Sensor 3 couples strongly to the secondary process (f2)
    s3 = 0.12 * f1 + 0.90 * f2 + np.random.normal(0, 0.25, n_samples)

    # Sensor 4 suffers an extreme electrical/instrumentation fault (massive localized white noise)
    s4 = 0.02 * f1 + 0.05 * f2 + np.random.normal(0, 1.40, n_samples)

    # 3. Compile into the final evaluation DataFrame
    df_asset = pd.DataFrame(
        data=np.vstack([s1, s2, s3, s4]).T,
        columns=['Sensor_1', 'Sensor_2', 'Sensor_3', 'Sensor_4']
    )
    # Instantiate Engine
    engine = DualSubspaceDiagnosticsEngine(df_asset)

    # Run PCA Engine (1x5 Dashboard)
    pca_metrics = engine.compute_empirical_pca(show_plot=True)

    # Run FA Engine (2x2 Dashboard, Retaining k=2 Common Physical Factors)
    fa_metrics = engine.compute_empirical_fa(k=2, show_plot=True)


--- Operationalizing Geometric De-correlation Layer (PCA) ---
Decomposing structural space of 4 features using 2500 samples.
Total System Variance (Trace[S]): 4.4674



--- Operationalizing Latent Subspace Layer (Factor Analysis) ---
Decomposing structural space of 4 sensors into 2 hidden physical factors.
Average System Communality (Shared Subspace Energy): 50.66%
Average System Uniqueness (Localized Noise Floor): 49.35%


### Solution for Question 1  

* **Part 1 (Physical Nature of Sensor 4):** A Uniqueness value ($\varphi^2$) approaching $100\%$ indicates that `Sensor_4` shares zero covariance with the rest of the array. It is completely decoupled from the real physical processes ($f_1, f_2$) driving the plant asset. Its data profile consists entirely of isolated white noise or severe electrical fault energy.
* **Part 2 (The PCA Loading Contamination):** Looking at the **Feature Loading Matrix** in the PCA dashboard, we see a striking anomaly: `PC 1` maps almost exclusively to a single variable—`Sensor_4` (indicated by the deep red block with a loading value near $1.0$). Because `Sensor_4` has a massive, unmitigated noise variance scale ($\sigma^2 \approx 2.0$), the PCA engine maximizes global variance by dedicating its entire primary axis (`PC 1`) to tracking this single noisy sensor. It completely isolates `Sensor_1`, `Sensor_2`, and `Sensor_3` to a loading of $0$ on `PC 1`.
* **Part 3 (Engineering Risk):** This presents a significant operational hazard. If you use standard PCA for monitoring, a single blown or short-circuited instrument will hijack your most critical monitoring axis (`PC 1`), inflating its eigenvalue to $\lambda_1 \approx 2.0$. This masks the true common underlying structural dynamics of the asset and renders your primary control chart useless for catching actual systemic failures.

### Solution for Question 2


* **Part 1 (Mathematical Jumble vs. Simple Structure):** Because PCA is bound to strict, orthogonal variance maximization, look at what happens to the remaining signals on `PC 2` and `PC 4`. `PC 2` and `PC 4` exhibit high loading values across *both* `Sensor_1` and `Sensor_2` simultaneously, while also bleeding into `Sensor_3`. This cross-loading represents a mathematically forced mixture of signals. Factor Analysis applies a **Varimax rotation** to eliminate this. By maximizing loading variance within columns, it completely isolates the signals: `Sensor_1` and `Sensor_2` group cleanly into `Factor 1`, while `Sensor_3` shifts entirely into `Factor 2`.
* **Part 2 (Operational Interpretability):** The PCA loading matrix forces a complex interpretation. If a diagnostic alarm trips on `PC 2`, an engineer has to figure out a messy combination of Sensors 1, 2, and 3. With the rotated FA framework, the layout is intuitive. `Factor 1` represents the isolated physical process causing redundancy in `Sensor_1` and `Sensor_2`, while `Factor 2` maps directly to the independent process monitored by `Sensor_3`. Troubleshooting root causes becomes instantaneous.


### Solution for Question 3

* **Part 1 (Curve Observations):** Moving from $k=1$ to $k=2$, the Mean $Q$ Statistic (SPE) plunges from $2.4$ down to $0.9$. When transitioning from $k=2$ to $k=3$, the curve flattens out into an explicit "elbow", decreasing at a much slower, linear rate down toward $0.1$.
* **Part 2 (The True Intrinsic Dimension):** Cross-referencing the $Q$ curve with the new loading matrix makes the selection clear. `PC 1` tracks the isolated noise of `Sensor_4`. `PC 2` captures the true, shared variance of the working sensors (`Sensor_1`, `Sensor_2`, `Sensor_3`). The sharp drop in the $Q$ statistic at $k=2$ tells us that once these two distinct behaviors are accounted for, the structured information is fully exhausted. Thus, $k=2$ represents the true hidden physical dimensionality of the asset.
* **Part 3 (The Danger of Over-fitting to $k=3$):** Look at `PC 3` in the loading matrix—it maps exclusively to the leftover residual variance of `Sensor_3`. Setting your truncation limit to $k=3$ pulls this residual space into your "clean" model subspace. This overfits the system, causing your Hotelling's $T^2$ control limits to trigger frequent false alarms due to minor, normal baseline variations.

### Solution for Question 4

* **The PCA Strategy:** Tracking the main principal subspace via $T^2$ alongside the residual space via the $Q$ statistic.
* *Vulnerability:* This approach is highly vulnerable to sensor degradation. As shown in the PCA loading matrix, if a single channel (`Sensor_4`) fails high, it structurally shifts the orientation of your principal axes, dragging the calculations for the remaining healthy channels along with it.


* **The FA Strategy:** Directly monitoring the rotated Latent Factor Scores.
* *Robustness:* This approach is highly robust. Because Factor Analysis explicitly separates localized noise via the diagonal uniqueness vector ($\Psi$), any single-channel sensor failure is safely absorbed by that channel's uniqueness profile ($\psi_4^2 \rightarrow 1$). The underlying latent factors (`Factor 1` and `Factor 2`) remain clean, stable, and uncorrupted, allowing the rest of your health monitoring network to function normally.